tests/test_database_types.py

```diff
@@ -178,60 +242,135 @@ def expand_params(testcase_func, param_num, param):
+        start = time.time()
+        already_seeded = self._create_table(src_conn, src_table, source_type)
+        if not already_seeded:
+            self._insert_to_table(
+                src_conn, src_table, source_type, enumerate(sample_values, 1)
+            )
+        insertion_source_duration = round(time.time() - start, 2)
 
-        self.table = TableSegment(self.src_conn, src_table_path, "id", None, ("col",), quote_columns=False)
-        self.table2 = TableSegment(self.dst_conn, dst_table_path, "id", None, ("col",), quote_columns=False)
+        start = time.time()
+        already_seeded = self._create_table(dst_conn, dst_table, target_type)
+        if not already_seeded:
+            values_in_source = src_conn.query(f"SELECT id, col FROM {src_table}", list)
+            self._insert_to_table(dst_conn, dst_table, target_type, values_in_source)
+        insertion_target_duration = round(time.time() - start, 2)
 
+        start = time.time()
         self.assertEqual(len(sample_values), self.table.count())
+        count1_duration = round(time.time() - start, 2)
+
+        start = time.time()
         self.assertEqual(len(sample_values), self.table2.count())
+        count2_duration = round(time.time() - start, 2)
+
+        # Ensure we actually checksum even for few samples!
+        ch_threshold = 5_000 if N_SAMPLES > DEFAULT_N_SAMPLES else 3
+        ch_factor = (
+            min(max(int(N_SAMPLES / 250_000), 2), 12)
+            if N_SAMPLES > DEFAULT_N_SAMPLES
+            else 2
+        )
+        ch_threads = 12
 
-        differ = TableDiffer(bisection_threshold=3, bisection_factor=2)  # ensure we actually checksum
+        start = time.time()
+        differ = TableDiffer(
+            bisection_threshold=ch_threshold,
+            bisection_factor=ch_factor,
+            max_threadpool_size=ch_threads,
+            debug=True,
+        )
         diff = list(differ.diff_tables(self.table, self.table2))
         expected = []
         self.assertEqual(expected, diff)
         self.assertEqual(0, differ.stats.get("rows_downloaded", 0))
+        checksum_duration = round(time.time() - start, 2)
 
+        start = time.time()
         # Ensure that Python agrees with the checksum!
-        differ = TableDiffer(bisection_threshold=1000000000)
+        dl_factor = int(N_SAMPLES / 100_000) if N_SAMPLES > DEFAULT_N_SAMPLES else 2
+        dl_threshold = int(N_SAMPLES / dl_factor) + 1
+        dl_threads = 8
+
+        differ = TableDiffer(
+            bisection_threshold=dl_threshold,
+            bisection_factor=dl_factor,
+            max_threadpool_size=dl_threads,
+        )
         diff = list(differ.diff_tables(self.table, self.table2))
         expected = []
         self.assertEqual(expected, diff)
         self.assertEqual(len(sample_values), differ.stats.get("rows_downloaded", 0))
+        download_duration = round(time.time() - start, 2)
+
+        print(
```

Comment:
I thought we agreed tests don't print. Use the logger.

Reply:
Yep, for sure, just a debugging relic! Will log to `debug`!